### PR TITLE
Rails 7.1

### PR DESCRIPTION
**Pull request by tbhi**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -25,7 +25,7 @@ gem 'pg'
 gem 'scenic'
 
 # Dashboard analytics
-gem 'energy-sparks_analytics', github: 'Energy-Sparks/energy-sparks_analytics', tag: '5.2.11'
+gem 'energy-sparks_analytics', github: 'Energy-Sparks/energy-sparks_analytics', branch: '20240617-rails71'
```
Comment (ldodds) [inline Gemfile:28]: reminder to switch to tagged version

File: config/application.rb [hunk 2]
```diff
@@ -12,6 +12,11 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.1
 
+    # Please, add to the `ignore` list any other `lib` subdirectories that do
+    # not contain `.rb` files, or that should not be reloaded or eager loaded.
+    # Common ones are `templates`, `generators`, or `middleware`, for example.
+    config.autoload_lib(ignore: %w(assets tasks))
```
Comment (ldodds) [inline config/application.rb:18]: this is set again below?

File: config/application.rb [hunk 3]
```diff
@@ -20,46 +25,29 @@ class Application < Rails::Application
...
-    #
-    # For now, explicitly ignore the generator from being auto-loaded.
-    Rails.autoloaders.main.ignore(Rails.root.join('lib/generators/comparison_report/comparison_report_generator.rb'))
-
-    #config.time_zone = 'London'
-    # optional - note it can be only :utc or :local (default is :utc)
-    # HAS to be UTC for group by date to work
-    config.active_record.default_timezone = :utc
-
+    config.eager_load_paths << Rails.root.join('app/models/areas')
```
Comment (ldodds) [inline config/application.rb:31]: is this necessary, there's no `app/models/areas` folder, just `area.rb`? Wondering if this is a legacy thing.

File: config/application.rb [hunk 4]
```diff
@@ -20,46 +25,29 @@ class Application < Rails::Application
...
-
     # session cookie config will be overridden in production.rb
     config.session_store :cookie_store, key: '_energy-sparks_session'
-
     config.after_initialize do
       if EnergySparks::FeatureFlags.active?(:use_site_settings_current_prices)
         BenchmarkMetrics.set_current_prices(prices: SiteSettings.current_prices)
       end
-
+      # https://stackoverflow.com/questions/77366033/allow-actiontext-tags-in-rails-7-1-with-new-sanitizers
```
Comment (ldodds) [inline config/application.rb:45]: what do we need to test to ensure this is working? Is it just image and other embeds in trix content?

File: config/environments/production.rb [hunk 5]
```diff
@@ -45,21 +44,31 @@
...
+
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
-  # config.force_ssl = true
+  config.force_ssl = true
 
-  # Include generic and useful information about system operation, but avoid logging too much
-  # information to avoid inadvertent exposure of personally identifiable information (PII).
-  config.log_level = :info
+  # Log to STDOUT by default
+  config.logger = ActiveSupport::Logger.new(STDOUT)
```
Comment (ldodds) [inline config/environments/production.rb:55]: do the logging changes in Rails 7.1 mean we no longer get log/production.log? I see mention of a broadcast logger.

File: app/models/dashboard_alert.rb [hunk 6]
```diff
@@ -33,6 +33,7 @@ class DashboardAlert < ApplicationRecord
   belongs_to :find_out_more, optional: true
   belongs_to :content_version, class_name: 'AlertTypeRatingContentVersion', foreign_key: :alert_type_rating_content_version_id
 
+  attribute :dashboard, :string
```
Comment (ldodds) [inline app/models/dashboard_alert.rb:36]: was this to fix a spec failure somewhere?

**Review by ldodds — APPROVED** [review]

File: config/environments/production.rb (hunk 5 above)
Comment (tbhi) [reply config/environments/production.rb:55]: looks like it
https://github.com/rails/rails/commit/2b1fa89e442ecb99e262a34eb11bc1110912cd49
https://github.com/rails/rails/issues/50369

File: config/application.rb (hunk 3 above)
Comment (tbhi) [reply config/application.rb]: removed

File: config/application.rb (hunk 4 above)
Comment (tbhi) [reply config/application.rb:43]: This was added in https://github.com/Energy-Sparks/energy-sparks/commit/e9c77fcdf96de1ad2f2325e37a43bd02d6cf4860#diff-50fc7bad8921b4b92c8b7baa63ae23af079068f87df641c94e6e7909b75deba0 - do charts relate to actiontext?

File: config/application.rb (hunk 2 above)
Comment (tbhi) [reply config/application.rb:18]: updated

File: config/environments/production.rb (hunk 5 above)
Comment (ldodds) [reply config/environments/production.rb:55]: ElasticBeanstalk has an option to request logs from web console either last 100 lines or full logs. Its expecting there to be a production.log so now fails:

```
2024/06/28 08:21:39.447424 [ERROR] An error occurred during execution of command [tail-log] - [GetTailLogs]. Stop running the command. Error: fail to check stat of file /var/app/containerfiles/logs/production.log: stat /var/app/containerfiles/logs/production.log: no such file or directory 
```

So to preserve this functionality we either need to find a workaround for EB or add the production.log back in

File: config/application.rb (hunk 4 above)
Comment (ldodds) [reply config/application.rb:44]: We embed charts in the "school specific description" fields for activities. This is a version of the instructions that includes charts that show the actual data for the school, so when they view the activity type, or their recorded activity, the charts are displayed. 

The charts are embedded in text using a trix extension. What's actually stored is a Mustache template (see `school_template_spec.rb`). But this gets processed to replace it with a chart tag. Looks like this config is avoiding ActionText stripping the attributes when the content is actually rendered. 

I've checked that these embedded charts are still rendering on test: https://test.energysparks.uk/schools/stepaside-primary-school/activities/2320

File: app/models/dashboard_alert.rb (hunk 6 above)
Comment (tbhi) [reply app/models/dashboard_alert.rb]: removed this one, only in app/models/staff_role.rb as rails complaining about non database back enum

**Pull request merged**